06x-new-kafka-roller.md

```diff
@@ -0,0 +1,331 @@
+  
+   - If `NON_ACTIVE_PURE_CONTROLLER` group is non empty, return the first node that can be restarted without impacting the quorum health (more on this later).
+   - If `ACTIVE_PURE_CONTROLLER` group is non empty, return the node if it can be restarted without impacting the quorum health. Otherwise return an empty set.
+   - If `COMBINED_AND_NOT_ACTIVE_CONTROLLER` group is non empty, return the first node that can be restarted without impacting the quorum health and the availability.
+   - If `COMBINED_AND_ACTIVE_CONTROLLER`  group is non empty, return the node if it can be restarted without impacting the quorum health and the availability. Otherwise return an empty set.
+   - If `BROKER` group is non empty, batch the broker nodes:
+       - build a map of nodes and their replicating partitions by sending describeTopics request to Admin API
+       - batch the nodes that do not have any partitions in common therefore can be restarted together
+       - remove nodes that have an impact on the availability from the batches (more on this later)
+       - return the largest batch
+   - If an empty batch is returned, that means none of the nodes met the safety conditions such as availability and qourum health impact. In this case, check their `numRetries` and if any of them is equal to or greater than `maxRetries`, throw `UnrestartableNodesException`. Otherwise increment their `numRetries` and repeat from step 2.
+
+11. Restart the nodes from the returned batch in parallel:
+   - If `numRestartAttempts` of a node is larger than `maxRestartAttempts`, throw `MaxRestartsExceededException`.
+   - Otherwise, restart each node and transition its state to `RESTARTED` and increment its `numRestartAttempts`.
+   - After restarting all the nodes in the batch, wait for their states to become `SERVING` until the configured `postOperationalTimeoutMs` is reached.
+   - If the timeout is reached, throw `TimeoutException`. If a node's `numRetries` is greater than or equal to `maxRetries`. Otherwise increment their `numRetries` and repeat from step 2.
+   - After all the nodes are in `SERVING` state, trigger preferred leader elections via Admin client. Wait for their states to become `LEADING_ALL_PREFERRED` until the configured `postOperationalTimeoutMs` is reached. If the timeout is reached, log a `WARN` message. 
+
+
+12. If there are no exceptions thrown at this point, the reconciliation completes successfully. If there were `UnrestartableNodesException`, `TimeoutException`, `MaxRestartsExceededException` or any other unexpected exceptions throws, the reconciliation fails.
```

Comment:
Hmm, seeing this algorithm, makes me think that it would be great to have a TLA+ expert, who would model a KafkaRoller and prove the correctness and safety of this algorithm formally using some safety and liveness properties (e.g., No node should exceed max retries (safety), no simultaneous restart of all controller nodes (safety) or All nodes should eventually be SERVING (liveness)) and more....

Reply:
@fvaleri might be able to help with this :)

Reply:
TLA+ is certainly a useful tool and this would be a great use case. The problem is that it has a steep learning curve and it requires a lot of practice to really get the benefit out of it. I can write a simple spec, but I lack the experience and time to write anything like this.

Things I don't like much about TLA+: the model you create is so distant from the actual implementation that it doesn't protect you from nasty bugs (see KRaft and Kafka replication protocol), it is really easy to hit the state space explosion so you have to know some mitigation techniques (still it may require big machines to execute the model over multiple days).